<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Wrapper to Test a Chromosome with LOCATER and kalis — TestLoci • locater</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Wrapper to Test a Chromosome with LOCATER and kalis — TestLoci"><meta name="description" content="A simple wrapper function designed to allow users to easliy run LOCATER in conjunction with kalis."><meta property="og:description" content="A simple wrapper function designed to allow users to easliy run LOCATER in conjunction with kalis."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">locater</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/simple_gwas_example.html">Simple LOCATER Example</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Wrapper to Test a Chromosome with LOCATER and kalis</h1>

      <div class="d-none name"><code>TestLoci.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A simple wrapper function designed to allow users to easliy run LOCATER in conjunction with kalis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TestLoci</span><span class="op">(</span></span>
<span>  <span class="va">y</span>,</span>
<span>  <span class="va">pars</span>,</span>
<span>  target.loci <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://kalis.louisaslett.com/reference/CacheSummary.html" class="external-link">L</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  A <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  test.opts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  num.ckpts <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  ckpt.first.locus <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use.forking <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  nthreads <span class="op">=</span> <span class="fl">1L</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>a <code>n</code> x <code>m</code> matrix of m quantitative phenotypes (one phenotype / column)</p></dd>


<dt id="arg-pars">pars<a class="anchor" aria-label="anchor" href="#arg-pars"></a></dt>
<dd><p>a <code>kalisParameters</code> object specifing HMM parameters to the Li and Stephens HMM as returned by <code><a href="https://kalis.louisaslett.com/reference/Parameters.html" class="external-link">kalis::Parameters</a></code></p></dd>


<dt id="arg-target-loci">target.loci<a class="anchor" aria-label="anchor" href="#arg-target-loci"></a></dt>
<dd><p>a vector of integers specifying index of variants among those cached with <code><a href="https://kalis.louisaslett.com/reference/CacheHaplotypes.html" class="external-link">kalis::CacheHaplotypes</a></code> to be screened with LOCATER</p></dd>


<dt id="arg-a">A<a class="anchor" aria-label="anchor" href="#arg-a"></a></dt>
<dd><p>a <code>n</code> x <code>q</code> matrix of q background covariates (an intercept – column of 1s – should be included). Just an intercept by default.</p></dd>


<dt id="arg-test-opts">test.opts<a class="anchor" aria-label="anchor" href="#arg-test-opts"></a></dt>
<dd><p>a list of testing options that can customized the behavior of <code>TestLoci</code> (further documentation coming)</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>a logical, if TRUE print progress statements as <code>TestLoci</code> progresses</p></dd>


<dt id="arg-num-ckpts">num.ckpts<a class="anchor" aria-label="anchor" href="#arg-num-ckpts"></a></dt>
<dd><p>an integer, the number of checkpoints that can be stored in memory. default = 0. Increasing even to just `2L` or `3L` yields significant speed ups.</p></dd>


<dt id="arg-ckpt-first-locus">ckpt.first.locus<a class="anchor" aria-label="anchor" href="#arg-ckpt-first-locus"></a></dt>
<dd><p>a logical, should the first target locus be stored as a checkpoint. Default is FALSE but if the first target locus is very far from start of the cached chromosome, TRUE may yield an acceleration.</p></dd>


<dt id="arg-use-forking">use.forking<a class="anchor" aria-label="anchor" href="#arg-use-forking"></a></dt>
<dd><p>a logical, is multiprocessing by process forking allowed?  Some relatively minor acclerations are possible if TRUE, but users should verify that it is safe to launch forked processes on their compute cluster.</p></dd>


<dt id="arg-nthreads">nthreads<a class="anchor" aria-label="anchor" href="#arg-nthreads"></a></dt>
<dd><p>an integer, number of threads available</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a data.table with one row per target locus including the following key fields</p><ul><li><p><code>locus.idx</code>: the index of corresponding variant (corresponding to <code>target.loci</code>)</p></li>
<li><p><code>phenotype</code>: the phenotype tested (given a column index of <code>y</code> )</p></li>
<li><p><code>tot</code>: LOCATER -log10 p-value</p></li>
<li><p><code>smt</code>: -log10 p-value from single marker testing</p></li>
<li><p><code>rd</code>: -log10 p-value returned by running Stable Distillation</p></li>
<li><p><code>qform</code>: -log10 p-value returned by quadratic form testing</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer></div>





  </body></html>

