<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simple LOCATER Example • locater</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simple LOCATER Example">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">locater</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/simple_gwas_example.html">Simple LOCATER Example</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simple LOCATER Example</h1>
            
      

      <div class="d-none name"><code>simple_gwas_example.Rmd</code></div>
    </div>

    
    
<p>This simple vignette runs LOCATER, with an API implemented in our
<code>R</code> package <code>locater</code>, on a toy example.</p>
<p>In the first section, we run a simple ancestry-based association test
with our a wrapper function <code>TestLoci</code>, which uses
<code>kalis</code> as the underlying ancestry inference engine. We take
a toy data set with 300 simulated haplotypes (150 samples) observed
across 400 loci using 20 simulated quantitative phenotypes where the
first phenotype is driven by 4 causal variants. The other 19 simulated
phenotypes are “null”: their mean is not a function of any of the
simulated genotypes. We then plot the results in a local hit-plot.</p>
<p>We then proceed in the second section to expose more of the
<code>locater</code> package API which allows users to pass clade
genotypes and local relatedness matrices inferred using their ancestry
inference engine of choice to <code>locater</code> testing functions. We
use this provide a simple demonstration of how <code>locater</code> may
be easily used to test output from any ancestry inference engine.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ryanchrist.github.io/locater/">locater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kalis.louisaslett.com/" class="external-link">kalis</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running in 64-bit mode using x86-64 architecture.</span></span>
<span><span class="co">#&gt; Loops unrolled to depth 4.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Currently not using any special instruction sets (WARNING: poor performance likely).</span></span>
<span><span class="co">#&gt; If this is unexpected (e.g. your CPU is Intel Haswell or newer architecture, or ARMv7+ with NEON support), then ensure that you are targeting the native architecture in compilation.  The easiest method is to add/change the following line in ~/.R/Makevars</span></span>
<span><span class="co">#&gt; CFLAGS=-march=native -mtune=native -O3</span></span></code></pre></div>
<div class="section level2">
<h2 id="boosting-smt-results-with-local-ancestry-inference-running-kalis-and-locater-together-with-easy-wrapper-function-testloci">Boosting SMT results with local ancestry inference: running kalis
and LOCATER together with easy wrapper function
<code>TestLoci</code><a class="anchor" aria-label="anchor" href="#boosting-smt-results-with-local-ancestry-inference-running-kalis-and-locater-together-with-easy-wrapper-function-testloci"></a>
</h2>
<p>We start by loading the haplotypes corresponding to our genomic
segment of interest.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load simulated haplotypes and recombination map</span></span>
<span><span class="fu"><a href="https://kalis.louisaslett.com/reference/CacheHaplotypes.html" class="external-link">CacheHaplotypes</a></span><span class="op">(</span><span class="va">SmallHaps</span><span class="op">)</span></span>
<span><span class="va">map</span> <span class="op">&lt;-</span> <span class="va">SmallMap</span></span></code></pre></div>
<p>Here we simulate a simple quantitative trait driven by four causal
variants.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate 20 Phenotypes where only the first is driven by local causal variants</span></span>
<span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://kalis.louisaslett.com/reference/CacheSummary.html" class="external-link">N</a></span><span class="op">(</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span> <span class="co"># number of samples</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fl">20</span> <span class="co"># number of phenotypes</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">n</span><span class="op">*</span><span class="va">m</span><span class="op">)</span>,<span class="va">n</span>,<span class="va">m</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://kalis.louisaslett.com/reference/QueryCache.html" class="external-link">QueryCache</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">204</span>,<span class="fl">205</span>,<span class="fl">207</span>,<span class="fl">210</span><span class="op">)</span><span class="op">)</span> <span class="co"># c(204,205,207) are indices of causal loci</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://kalis.louisaslett.com/reference/CacheSummary.html" class="external-link">N</a></span><span class="op">(</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span> <span class="op">+</span> <span class="va">g</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fu"><a href="https://kalis.louisaslett.com/reference/CacheSummary.html" class="external-link">N</a></span><span class="op">(</span><span class="op">)</span>,<span class="fl">2</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">y</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">150</span>,<span class="fl">10</span>,<span class="fl">30</span>,<span class="fl">100</span><span class="op">)</span><span class="co">#c(150,8,20)</span></span>
<span>                         <span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">A</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span>,<span class="va">n</span>,<span class="fl">1</span><span class="op">)</span> <span class="co"># background covariate matrix is just an index here</span></span></code></pre></div>
<p>Now we determine target variants based on initial SMT results.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Determine target loci based on SMT</span></span>
<span><span class="va">smt.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestCachedMarkers.html">TestCachedMarkers</a></span><span class="op">(</span><span class="va">y</span>, A <span class="op">=</span> <span class="va">A</span><span class="op">)</span></span>
<span><span class="va">target.loci</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FindTargetVars.html">FindTargetVars</a></span><span class="op">(</span><span class="va">map</span>, min.cM <span class="op">=</span> <span class="fl">0.1</span>, initial.targets <span class="op">=</span> <span class="va">smt.res</span>, smt.thresh <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p>We iterate LOCATER over the target loci.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Specify kalis parameters </span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://kalis.louisaslett.com/reference/Parameters.html" class="external-link">Parameters</a></span><span class="op">(</span><span class="fu"><a href="https://kalis.louisaslett.com/reference/CalcRho.html" class="external-link">CalcRho</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span>,s <span class="op">=</span> <span class="fl">0.0001</span><span class="op">)</span>,mu <span class="op">=</span> <span class="fl">1e-16</span><span class="op">)</span> <span class="co"># specify HMM parameters</span></span>
<span></span>
<span><span class="co"># Specify a list of optional arguments to control exactly how locater tests each position. If more than one combination of testing parameters is desired, a data.frame with multiple rows can be given and locater will efficiently evaluate all testing options.</span></span>
<span><span class="va">test.opts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"max.k"</span> <span class="op">=</span> <span class="fl">128</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#set.seed(53)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestLoci.html">TestLoci</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">pars</span>, target.loci <span class="op">=</span> <span class="va">target.loci</span>, test.opts <span class="op">=</span> <span class="va">test.opts</span><span class="op">)</span></span></code></pre></div>
<p>Then we plot our results and compare to a SMT in a local Manhattan
plot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">smt.p1</span> <span class="op">&lt;-</span> <span class="va">smt.res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">res.p1</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="va">phenotype</span><span class="op">==</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"locus.idx"</span>,<span class="st">"tot"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="va">COLORS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palette.html" class="external-link">palette.colors</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>, type <span class="op">=</span> <span class="st">"n"</span>, </span>
<span>     ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">smt.p1</span>,<span class="va">res.p1</span><span class="op">$</span><span class="va">tot</span><span class="op">)</span>,na.rm<span class="op">=</span><span class="cn">T</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span><span class="op">)</span>,</span>
<span>     las<span class="op">=</span><span class="fl">1</span>,bty<span class="op">=</span><span class="st">"n"</span>,</span>
<span>     ylab <span class="op">=</span> <span class="st">"-log10 p-value"</span>, xlab <span class="op">=</span> <span class="st">"cM position"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">map</span>,<span class="va">smt.p1</span>,col<span class="op">=</span><span class="va">COLORS</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,pch<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span><span class="va">map</span><span class="op">[</span><span class="va">res.p1</span><span class="op">$</span><span class="va">locus.idx</span><span class="op">]</span>,<span class="va">res.p1</span><span class="op">$</span><span class="va">tot</span>,col<span class="op">=</span><span class="va">COLORS</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,pch<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>,legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SMT"</span>,<span class="st">"LOCATER"</span><span class="op">)</span>,</span>
<span>       fill <span class="op">=</span> <span class="va">COLORS</span>, bty <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_gwas_example_files/figure-html/visualize_manhattan-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="using-other-ancestry-inference-engines">Using other ancestry inference engines<a class="anchor" aria-label="anchor" href="#using-other-ancestry-inference-engines"></a>
</h2>
<p>The <code>locater</code> package API exposes all of the sub-routines
needed to for running LOCATER in combination with any ancestry inference
engine. The <code>TestLoci</code> function in the <code>locater</code>
package includes the following function calls.</p>
<div class="section level3">
<h3 id="testing-inferred-discrete-clade-genotypes">Testing inferred discrete clade genotypes<a class="anchor" aria-label="anchor" href="#testing-inferred-discrete-clade-genotypes"></a>
</h3>
<p>Building on our example above, we can test inferred discrete clade
genotypes with Stable Distillation with a call to the
<code>distill_pivot_par</code> function. Let <code>y</code> be a
<code>n \times m</code> matrix of phenotypes, <code>x</code> be a
<code>n \times p</code> matrix of predictors (inferred clade genotypes),
and <code>Q</code> be the matrix from the QR decomposition of the
<code>n \times q</code> background covariate matrix <code>A = QR</code>.
By default, we request that the Renyi Outlier Test search for at most 16
distinct active predictors among the colums of <code>x</code>. Below we
start by simulating a set of inferred clade genotypes in order to run
the procedure, but we assume that those would come from an ancestry
inference engine. Among theose simulated genotypes, we include the 4
genotypes from above (in matrix <code>g</code>) that actually affect the
first phenotype.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fl">50</span> <span class="co"># number of inferred genotypes</span></span>
<span><span class="va">num_true_causal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span>,<span class="fu">Matrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/sparseMatrix.html" class="external-link">sparseMatrix</a></span><span class="op">(</span>i <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample.int</a></span><span class="op">(</span><span class="va">n</span>,<span class="fl">10</span><span class="op">*</span><span class="op">(</span><span class="va">p</span><span class="op">-</span><span class="va">num_true_causal</span><span class="op">)</span>,replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>                          j <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">p</span><span class="op">-</span><span class="va">num_true_causal</span><span class="op">)</span>, each<span class="op">=</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>                          x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">8</span><span class="op">)</span>,<span class="fl">2</span>,<span class="fl">2</span><span class="op">)</span>, <span class="op">(</span><span class="va">p</span><span class="op">-</span><span class="va">num_true_causal</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                          dims <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">n</span>,<span class="va">p</span><span class="op">-</span><span class="va">num_true_causal</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">Q</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/qraux.html" class="external-link">qr.Q</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/qr.html" class="external-link">qr</a></span><span class="op">(</span><span class="va">A</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sd_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/distill_pivot_par.html">distill_pivot_par</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">x</span>, <span class="va">Q</span>, max_num_causal <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span></code></pre></div>
<p>Below we plot the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-log_{10}</annotation></semantics></math>
p-value obtained of running parallelized Stable Distillation for each of
our 20 simulated phenotypes as a bar plot.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">sd_res</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>,names.arg<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>,xlab<span class="op">=</span><span class="st">"Phenotype"</span>,ylab<span class="op">=</span><span class="st">"-log10 SD p-value"</span>,las<span class="op">=</span><span class="fl">1</span>,cex.names<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_gwas_example_files/figure-html/visualize_sd_results-1.png" width="700"></p>
<p>Only the first phenotype, the one driven by our simulated causal
variants, has a very significant SD p-value. <code>sd_res</code> will
contain a list with lots of other information alongside <code>m</code>
p-values plotted above. More information about this function is
available on it’s manual page, accessible by typing
<code><a href="../reference/distill_pivot_par.html">?locater::distill_pivot_par</a></code> into <code>R</code>.</p>
</div>
<div class="section level3">
<h3 id="testing-inferred-local-relatedness-matrices">Testing inferred local relatedness matrices<a class="anchor" aria-label="anchor" href="#testing-inferred-local-relatedness-matrices"></a>
</h3>
<p>Building on example of testing inferred discrete clade genotypes
above, we can test inferred local relatedness matrices with the
<code>TestCladeMat</code> function in the <code>locater</code> package.
We assume that we have a inferred local relatedness matrix
<code>M</code> returned by some ancestry inference engine. Here, we just
construct a simple <code>M</code> from the genotypes simulated for our
Stable Distillation example above. Note, <code>M</code> must be passed
to <code>TestCladeMat</code> as a base <code>R</code>
<code>matrix</code>. As above, we account for the background covaraites
via <code>Q</code>. The argument <code>k</code> specifies the maximum
number of eigenvalues of <code>(I-QQ')M(I-QQ')</code> that we will
budget in hopes of gaining precise p-value estimates.
<code>TestCladeMat</code> allows for much finer control of this process;
please see the documentation by typing
<code><a href="../reference/TestCladeMat.html">?locater::TestCladeMat</a></code> into <code>R</code> for more
details. Here we start by setting <code>k=0</code> which means
<code>TestCladeMat</code> will just return the Satterthwaite
p-values.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">M</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/crossprod.html" class="external-link">tcrossprod</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">q_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestCladeMat.html">TestCladeMat</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">M</span>, <span class="va">Q</span>, k <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><code>TestCladeMat</code> returns a <code>data.frame</code> with
<code>m</code> rows where row <code>i</code> corresponds to the results
for phenotype <code>y[,i]</code>. As above, we can plot the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-log_{10}</annotation></semantics></math>
p-values returned by our quadratic form testing procedure. Again, we see
an association with the first (non-null) phenotype.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">q_res</span><span class="op">$</span><span class="va">qform</span>,names.arg<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>,xlab<span class="op">=</span><span class="st">"Phenotype"</span>,ylab<span class="op">=</span><span class="st">"-log10 Quadratic Form p-value"</span>,las<span class="op">=</span><span class="fl">1</span>,cex.names<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_gwas_example_files/figure-html/visualize_qform_results-1.png" width="700">
However, these p-value estimates are not precise/reliable due to the use
of the Satterthwaite approximation for the null distribution.
<code>TestCladeMat</code> warns of this fact in the returned results by
setting <code>precise=FALSE</code> .</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">q_res</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;        qform        obs obs.T precise exit.status</span></span>
<span><span class="co">#&gt; 1 74.8638206 12024.8522    NA   FALSE           0</span></span>
<span><span class="co">#&gt; 2  0.7115142  1021.7774    NA   FALSE           0</span></span>
<span><span class="co">#&gt; 3  0.1840489   746.9028    NA   FALSE           0</span></span>
<span><span class="co">#&gt; 4  0.1064964   674.4044    NA   FALSE           0</span></span>
<span><span class="co">#&gt; 5  0.1059957   673.8431    NA   FALSE           0</span></span></code></pre></div>
<p>So we may have used this crude approximation in a first pass screen
through the genome. In order to obtain reliable estimates using the
novel tail approximation approaches proposed in our paper, we can test
this locus again, allowing for up to <code>k</code> eigenvalues to be
evaluated.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">q_res_precise</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TestCladeMat.html">TestCladeMat</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">M</span>, <span class="va">Q</span>, k <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>Plotting these results, we see a significant but much more modest
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mo>−</mo><mi>l</mi><mi>o</mi><msub><mi>g</mi><mn>10</mn></msub></mrow><annotation encoding="application/x-tex">-log_{10}</annotation></semantics></math>
p-value estimate associated with the first phenotype.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">q_res_precise</span><span class="op">$</span><span class="va">qform</span>,names.arg<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">m</span>,xlab<span class="op">=</span><span class="st">"Phenotype"</span>,ylab<span class="op">=</span><span class="st">"-log10 Quadratic Form p-value"</span>,las<span class="op">=</span><span class="fl">1</span>,cex.names<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span></span></code></pre></div>
<p><img src="simple_gwas_example_files/figure-html/visualize_qform_results_precise-1.png" width="700"></p>
<p>Printing, the results, we see that <code>TestCladeMat</code> says
these results/p-values are reliable (<code>precise = TRUE</code>).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">q_res_precise</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;        qform        obs obs.T precise exit.status</span></span>
<span><span class="co">#&gt; 1 30.8803872 12024.8522    NA    TRUE           0</span></span>
<span><span class="co">#&gt; 2  0.7335176  1021.7774    NA    TRUE           0</span></span>
<span><span class="co">#&gt; 3  0.1895087   746.9028    NA    TRUE           0</span></span>
<span><span class="co">#&gt; 4  0.1047127   674.4044    NA    TRUE           0</span></span>
<span><span class="co">#&gt; 5  0.1041659   673.8431    NA    TRUE           0</span></span></code></pre></div>
<p>We see that the precise association signal for phenotype 1 found here
by quadratic form testing is now about as strong as the one observed
using Stable Distillation. Note, the number of null genotypes we
simulated here is small: we set <code>p=ncol(x)=50</code>. If we were to
consider a relatedness matrix with many more clades, we would see the
signal found by quadratic form testing drop much more quickly than the
signal found via Stable Distillation (which is much for robust to the
inclusion of null predictors/genotypes).</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Christ.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
